pig/src/main/java/com/netflix/iceberg/pig/IcebergStorage.java

```diff
@@ -0,0 +1,313 @@
+  }
+
+  private com.netflix.iceberg.expressions.Expression convert(OpType op, Column col, Const constant) {
+    String name = col.getName();
+    Object value = constant.getValue();
+
+    switch (op) {
+      case OP_GE: return greaterThanOrEqual(name, value);
+      case OP_GT: return greaterThan(name, value);
+      case OP_LE: return lessThanOrEqual(name, value);
+      case OP_LT: return lessThan(name, value);
+      case OP_EQ: return equal(name, value);
+      case OP_NE: return notEqual(name, value);
+    }
+
+    throw new RuntimeException(format("[%s]: Failed to pushdown expression: %s %s %s", signature, col, op, constant));
+  }
+
+  @Override
+  public List<OperatorSet> getFeatures() {
+    return Collections.singletonList(OperatorSet.PROJECTION);
+  }
+
+  @Override
+  public RequiredFieldResponse pushProjection(RequiredFieldList requiredFieldList) {
+    LOG.info(format("[%s]: pushProjection() -> %s", signature, requiredFieldList));
+
+    try {
+      List<String> projection = requiredFieldList.getFields().stream().map(RequiredField::getAlias).collect(Collectors.toList());
+
+      storeInUDFContext(ICEBERG_PROJECTED_FIELDS, (Serializable) projection);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return new RequiredFieldResponse(true);
+  }
+
+  @Override
+  public void setUDFContextSignature(String signature) {
+    this.signature = signature;
+  }
+
+  private void storeInUDFContext(String key, Serializable value) throws IOException {
+    Properties properties = UDFContext.getUDFContext().getUDFProperties(this.getClass(), new String[]{signature});
+
+    properties.setProperty(key, ObjectSerializer.serialize(value));
+  }
+
+  private void copyUDFContextToConfiguration(Configuration conf, String key) {
+    String value = UDFContext.getUDFContext().getUDFProperties(this.getClass(), new String[]{signature}).getProperty(key);
+
+    if (value != null) {
+      conf.set(key, value);
+    }
+  }
+
+  @Override
+  public String relativeToAbsolutePath(String location, Path curDir) throws IOException {
+    return location;
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends Serializable> T getFromUDFContext(String key, Class<T> clazz) throws IOException {
+    Properties properties = UDFContext.getUDFContext().getUDFProperties(this.getClass(), new String[]{signature});
+
+    return (T) ObjectSerializer.deserialize(properties.getProperty(key));
+  }
+
+  private Table load(String location, Job job) throws IOException {
+    Class<?> tablesImpl = job.getConfiguration().getClass(PIG_ICEBERG_TABLES_IMPL, HadoopTables.class);
+
+    Table result = tables.get(location);
+
+    if (result == null) {
+      try {
+        LOG.info(format("[%s]: Loading table from location: %s using impl: %s", signature, location, tablesImpl));
+        Tables iceberg = (Tables) ReflectionUtils.newInstance(tablesImpl, job.getConfiguration());
```

Comment:
Could you separate out the reflection parts so that it reuses the same Tables instance each time? Then this would be able to use a single metastore connection instead of opening a new one for every load. Or is this called only once?

Reply:
No, you're right.  This will create a new instance of the tables impl for each location.  Will update.